### PR TITLE
Rename BartForMaskedLM -> BartForConditionalGeneration

### PR DESCRIPTION
**Pull request by sshleifer**

### CODE REVIEW

**Review by thomwolf — CHANGES_REQUESTED** [review]
Up to now, model classes have been named from the generic task they can handle (ex. `BertForSequenceClassification`  instead of `BertForNLI`)

We should do the same here and keep task specific names for `pipelines` which are intended for this.

IMO the generic name here is something like `BartForConditionalGeneration`

**Comment by sshleifer** [comment]
Docs:
image

File: tests/test_modeling_bart.py [hunk 1]
```diff
@@ -243,15 +245,15 @@ def test_lm_uneven_forward(self):
             decoder_ffn_dim=32,
             max_position_embeddings=48,
         )
-        lm_model = BartForMaskedLM(config)
-        context = torch.Tensor([[71, 82, 18, 33, 46, 91, 2], [68, 34, 26, 58, 30, 2, 1]]).long()
-        summary = torch.Tensor([[82, 71, 82, 18, 2], [58, 68, 2, 1, 1]]).long()
-        logits, enc_features = lm_model.forward(input_ids=context, decoder_input_ids=summary)
+        lm_model = BartForConditionalGeneration(config).to(torch_device)
```
Comment (sshleifer) [inline tests/test_modeling_bart.py]: This fix should be in a separate PR

File: docs/source/model_doc/bart.rst [hunk 2]
```diff
@@ -18,26 +18,28 @@ The Authors' code can be found `here <https://github.com/pytorch/fairseq/tree/ma
...
 Implementation Notes
 ~~~~~~~~~~~~~~~~~~~~
-- Bart doesn't use :obj:`token_type_ids`, for sequence classification just use BartTokenizer.encode to get the proper splitting.
-- Inputs to the decoder are created by BartModel.forward if they are not passed. This is different than some other model APIs.
-- Model predictions are intended to be identical to the original implementation. This only works, however, if the string you pass to fairseq.encode starts with a space.
-- Decoder inputs are created automatically by the helper function ``transformers.modeling_bart._prepare_bart_decoder_inputs``
-BartModel
-- ``MaskedLM.generate`` should be used for summarization, see the example in that docstrings
+- Bart doesn't use :obj:`token_type_ids` for sequence classification. Use BartTokenizer.encode to get the proper splitting.
+- ``BartModel.forward`` will create decoder inputs (using the helper function ``transformers.modeling_bart._prepare_bart_decoder_inputs``)  if they are not passed. This is different than some other modeling APIs.
```
Comment (thomwolf) [inline docs/source/model_doc/bart.rst]: "The forward pass of `BartModel`" we don't want to encourage people to use `model.forward()`, it's a bad PyTorch practice.

File: docs/source/model_doc/bart.rst [hunk 3]
```diff
@@ -18,26 +18,28 @@ The Authors' code can be found `here <https://github.com/pytorch/fairseq/tree/ma
...
+- Model predictions are intended to be identical to the original implementation. This only works, however, if the string you pass to ``fairseq.encode`` starts with a space.
+- ``BartForConditionalGeneration.generate`` should be used for summarization, see the example in that docstrings
```
Comment (thomwolf) [inline docs/source/model_doc/bart.rst]: "for conditional generation tasks like summarization"

File: src/transformers/__init__.py [hunk 4]
```diff
@@ -206,7 +206,12 @@
         XLMForQuestionAnsweringSimple,
         XLM_PRETRAINED_MODEL_ARCHIVE_MAP,
     )
-    from .modeling_bart import BartForSequenceClassification, BartModel, BartForMaskedLM
+    from .modeling_bart import (
+        BartForSequenceClassification,
+        BartModel,
+        BartForConditionalGeneration,
+        BartForConditionalGeneration,
```
Comment (thomwolf) [inline src/transformers/__init__.py]: duplicate

File: src/transformers/convert_bart_original_pytorch_checkpoint_to_pytorch.py [hunk 5]
```diff
@@ -87,7 +93,7 @@ def convert_bart_checkpoint(checkpoint_path, pytorch_dump_folder_path):
     # Check results
 
     if checkpoint_path == "bart.large.cnn":  # generate doesnt work yet
```
Comment (thomwolf) [inline src/transformers/convert_bart_original_pytorch_checkpoint_to_pytorch.py]: Remove this "generate doesnt work yet" comment?

File: src/transformers/convert_bart_original_pytorch_checkpoint_to_pytorch.py [hunk 6]
```diff
@@ -87,7 +93,7 @@ def convert_bart_checkpoint(checkpoint_path, pytorch_dump_folder_path):
...
-        model = BartForMaskedLM(config, base_model=model)
+        model = BartForConditionalGeneration(config, base_model=model)
```
Comment (thomwolf) [inline src/transformers/convert_bart_original_pytorch_checkpoint_to_pytorch.py]: Can you also fix all the `model.forward()` calls in this script?

**Review by thomwolf — APPROVED** [review]
Ok LGTM, thanks @sshleifer

**Pull request merged**